addons/account/models/company.py

```diff
@@ -642,57 +645,55 @@ def _check_hash_integrity(self):
             # We need the `sudo()` to ensure that all the moves are searched, no matter the user's access rights.
             # This is required in order to generate consistent hashes.
             # It is not an issue, since the data is only used to compute a hash and not to return the actual values.
-            field_names = [
-                *self.env['account.move']._get_integrity_hash_fields(),
-                'inalterable_hash',
-                'journal_id',
-                'move_type',
-                'sequence_prefix',
-                'sequence_number',
-                'secure_sequence_number',
-            ]
-            moves = self.env['account.move'].sudo().search_fetch(
+            self.env['account.move'].flush_model()
+            query = self.env['account.move'].sudo()._search(
                 domain=[
                     ('journal_id', '=', journal.id),
                     ('inalterable_hash', '!=', False),
                 ],
-                field_names=field_names,
-                order="secure_sequence_number ASC NULLS LAST, sequence_prefix, sequence_number ASC"
+                order="secure_sequence_number ASC NULLS LAST, sequence_prefix, sequence_number ASC",
             )
-
-            if not moves:
-                results.append({
-                    'journal_name': journal.name,
-                    'restricted_by_hash_table': 'V',
-                    'status': 'no_data',
-                    'msg_cover': _('There is no journal entry flagged for accounting data inalterability yet.'),
-                })
-                continue
-
+            self.env.cr.execute(query.select())
             prefix2result = defaultdict(lambda: {
                 'first_move': self.env['account.move'],
                 'last_move': self.env['account.move'],
                 'corrupted_move': self.env['account.move'],
             })
             last_move = self.env['account.move']
-            current_hash_version = 1
-            for move in moves:
-                prefix_result = prefix2result[move.sequence_prefix]
-                if prefix_result['corrupted_move']:
+            stop = False
+            while not stop:
+                move_ids = self.env.cr.fetchmany(INTEGRITY_HASH_BATCH_SIZE)
```

Comment:
My bad, after some investigations, `fetchmany` is actually too risky to use because the ORM can fire queries at any time. (then messing the state of the cursor :boom: ) To reproduce, just put a `self.env.invalidate_all()` in the tests.
We need to use server side cursors like this intead. We only need to take care of closing it when we are done with it. (ideally in a `try/finally` but it should be good enough like suggested)
Also, if you use `execute_query`, you don't need to flush manually anymore :+1: 

```diff
diff --git a/addons/account/models/company.py b/addons/account/models/company.py
index 83b027591a37..8d1522c69b43 100644
--- a/addons/account/models/company.py
+++ b/addons/account/models/company.py
@@ -7,6 +7,7 @@ import calendar
 from odoo import fields, models, api, _, Command
 from odoo.exceptions import ValidationError, UserError, RedirectWarning
 from odoo.osv import expression
+from odoo.tools import SQL
 from odoo.tools.mail import is_html_empty
 from odoo.tools.misc import format_date
 from odoo.tools.float_utils import float_round, float_is_zero
@@ -645,7 +646,6 @@ class ResCompany(models.Model):
             # We need the `sudo()` to ensure that all the moves are searched, no matter the user's access rights.
             # This is required in order to generate consistent hashes.
             # It is not an issue, since the data is only used to compute a hash and not to return the actual values.
-            self.env['account.move'].flush_model()
             query = self.env['account.move'].sudo()._search(
                 domain=[
                     ('journal_id', '=', journal.id),
@@ -653,19 +653,15 @@ class ResCompany(models.Model):
                 ],
                 order="secure_sequence_number ASC NULLS LAST, sequence_prefix, sequence_number ASC",
             )
-            self.env.cr.execute(query.select())
             prefix2result = defaultdict(lambda: {
                 'first_move': self.env['account.move'],
                 'last_move': self.env['account.move'],
                 'corrupted_move': self.env['account.move'],
             })
             last_move = self.env['account.move']
-            stop = False
-            while not stop:
-                move_ids = self.env.cr.fetchmany(INTEGRITY_HASH_BATCH_SIZE)
+            self.env.execute_query(SQL("DECLARE hashed_moves CURSOR FOR %s", query.select()))
+            while move_ids := self.env.execute_query(SQL("FETCH %s FROM hashed_moves", INTEGRITY_HASH_BATCH_SIZE)):
                 moves = self.env['account.move'].browse(move_id[0] for move_id in move_ids)
-                if len(move_ids) < INTEGRITY_HASH_BATCH_SIZE:
-                    stop = True

                 if not moves and not last_move:
                     results.append({
@@ -695,6 +691,8 @@ class ResCompany(models.Model):
                     prefix_result['last_move'] = move
                     last_move = move

+            self.env.execute_query(SQL("CLOSE hashed_moves"))
+
             for prefix, prefix_result in prefix2result.items():
                 if corrupted_move := prefix_result['corrupted_move']:
                     results.append({
```